### DISCUSSION TITLE
Purpose of m_undrrelxr in AMReX_MLABecLaplacian.H

### DISCUSSION THREAD

**Discussion by ankithadas** [AMReX-Codes/amrex#3530] [Q&A] [answered]
Hey AMReX team

I quite didn't understand the purpose of the variable m_undrrelxr in AMReX_MLCellLinOp.H. I can see it is used in abec_gsrb in MLABecLaplacian but I couldn't figure out where the values inside m_undrrelxr are assigned. I believe m_undrrelxr is initialized to quite_NaNs. My question is what values does this variable store and why is it used in the Gauss Seidel RB solver ? Any help would be much appreciated

Kind regards,
Ankith

**Comment by WeiqunZhang** [comment, marked as answer]
`m_undrrelxr` is initialized here https://github.com/AMReX-Codes/amrex/blob/f3598d2f7b6beca566e19d918b6e9976f8ed5a88/Src/LinearSolvers/MLMG/AMReX_MLCellLinOp.H#L1578, where it calls functions like  https://github.com/AMReX-Codes/amrex/blob/f3598d2f7b6beca566e19d918b6e9976f8ed5a88/Src/LinearSolvers/MLMG/AMReX_MLLinOp_K.H#L377.

It is used for convenience. For cells near the physical domain boundary or coarse/fine boundary, the stencil there is different. A normal 1d Laplacian stencil might be something like `(1, -2, 1)`. But if the left point is outside a homogeneous Neumann boundary, the stencil becomes something like `(N/A, -1, 1)`. In GSRB, `-1` instead of `-2` should be the diagonal component of the matrix. `m_underrelxr` is there to help us to go from `-2` to `-1` (or something else depending the boundary condition).
